src/node_watchdog.cc

```diff
@@ -39,66 +39,107 @@ using v8::Local;
 using v8::Object;
 using v8::Value;
 
-Watchdog::Watchdog(v8::Isolate* isolate, uint64_t ms, bool* timed_out)
-    : isolate_(isolate), timed_out_(timed_out) {
+void WatchdogService::Init() {
+  if (!initialized_) {
+    initialized_ = true;
+    sleep_until_hrtime_ = std::numeric_limits<uint64_t>::max();
 
-  int rc;
-  rc = uv_loop_init(&loop_);
-  if (rc != 0) {
-    OnFatalError("node::Watchdog::Watchdog()", "Failed to initialize uv loop.");
+    int rc = uv_mutex_init(&mutex_);
+    CHECK_EQ(0, rc);
+
+    rc = uv_cond_init(&cond_);
+    CHECK_EQ(0, rc);
+
+    rc = uv_thread_create(&thread_, &WatchdogService::Run, this);
+    CHECK_EQ(0, rc);
   }
+}
 
-  rc = uv_async_init(&loop_, &async_, [](uv_async_t* signal) {
-    Watchdog* w = ContainerOf(&Watchdog::async_, signal);
-    uv_stop(&w->loop_);
-  });
+void WatchdogService::Run(void* arg) {
+  WatchdogService* ws = static_cast<WatchdogService*>(arg);
 
-  CHECK_EQ(0, rc);
+  while (true) {
+    uv_mutex_lock(&ws->mutex_);
 
-  rc = uv_timer_init(&loop_, &timer_);
-  CHECK_EQ(0, rc);
+    const uint64_t hrtime = uv_hrtime();
+    while (!ws->watchdog_timers_.empty() &&
+           ws->watchdog_timers_.begin()->first <= hrtime) {
+      auto element = ws->watchdog_timers_.begin();
+      WatchdogTimer* watchdog_timer = element->second;
+      ws->watchdog_timers_.erase(element);
 
-  rc = uv_timer_start(&timer_, &Watchdog::Timer, ms, 0);
-  CHECK_EQ(0, rc);
+      *watchdog_timer->timed_out = true;
+      watchdog_timer->isolate->TerminateExecution();
+    }
 
-  rc = uv_thread_create(&thread_, &Watchdog::Run, this);
-  CHECK_EQ(0, rc);
-}
+    uv_mutex_unlock(&ws->mutex_);
 
+    if (ws->should_destroy_) break;
 
-Watchdog::~Watchdog() {
-  uv_async_send(&async_);
-  uv_thread_join(&thread_);
+    uv_mutex_lock(&ws->mutex_);
+    ws->sleep_until_hrtime_ = std::numeric_limits<uint64_t>::max();
+    if (!ws->watchdog_timers_.empty()) {
+      ws->sleep_until_hrtime_ = ws->watchdog_timers_.begin()->first;
+    }
 
-  uv_close(reinterpret_cast<uv_handle_t*>(&async_), nullptr);
+    uint64_t sleep_nanos = ws->sleep_until_hrtime_ - hrtime;
+    uv_cond_timedwait(&ws->cond_, &ws->mutex_, sleep_nanos);
+    uv_mutex_unlock(&ws->mutex_);
+  }
```

Comment:
The logic here doesn't look quite right to me. It should enter the loop holding the lock and not drop it until it breaks out of the loop again.

Reply:
Quite right you are.